### PR TITLE
[SPARK-11625][SQL] add java test for typed aggregate

### PR DESCRIPTION
**Pull request by cloud-fan**

### CODE REVIEW

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java [hunk 1]
```diff
@@ -366,4 +368,52 @@ public void testNestedTupleEncoder() {
...
+      },
+      e.STRING());
+
+    Dataset<Tuple2<String, Integer>> agged = grouped.agg(new IntSumOf().toColumn(e.INT(), e.INT()));
+    Assert.assertEquals(Arrays.asList(tuple2("a", 3), tuple2("b", 3)), agged.collectAsList());
+
+    Dataset<Tuple4<String, Integer, Long, Long>> agged2 = grouped.agg(
+      new IntSumOf().toColumn(e.INT(), e.INT()),
+      (TypedColumn<Tuple2<String, Integer>, Long>) (Object) expr("sum(_2)").as(e.LONG()),
+      (TypedColumn<Tuple2<String, Integer>, Long>) (Object) count("*"));
```
Comment (cloud-fan) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java]: java doesn't support contravariant in `TypedColumn`, which makes this typed aggregate API hard to use at java side.

cc @marmbrus @rxin

Comment (marmbrus) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java]: Yeah, this is unfortunate.  I think that the only thing we can do is use the `DataFrame` version of `agg` and then cast afterwards.

```java
    Dataset<Tuple4<String, Integer, Long, Long>> agged2 = grouped.agg(
            new IntSumOf().toColumn(e.INT(), e.INT()),
            expr("sum(_2)"),
            count("*")).as(e.tuple(e.STRING(), e.INT(), e.LONG(), e.LONG()));
```

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java [hunk 2]
```diff
@@ -34,6 +34,8 @@
 import org.apache.spark.sql.catalyst.encoders.Encoder$;
```
Comment (marmbrus) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java]: We should do something about this.  I think that if we rename the object to `Encoders` then Scala will build static forwarders and we can just call `Encoders.STRING()`

**Comment by cloud-fan** [comment]
retest this please.

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please.

**Comment by cloud-fan** [comment]
retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/GroupedDataset.scala [hunk 3]
```diff
@@ -137,8 +137,19 @@ class GroupedDataset[K, T] private[sql](
   }
 
   // To ensure valid overloading.
-  protected def agg(expr: Column, exprs: Column*): DataFrame =
-    groupedData.agg(expr, exprs: _*)
+  @scala.annotation.varargs
+  def agg(expr: Column, exprs: Column*): DataFrame =
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/GroupedDataset.scala]: Maybe we should just move this to `Queryable` with docs.

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/GroupedDataset.scala]: Oh sorry, maybe we need to make `GroupQueryable`?

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/GroupedDataset.scala]: @rxin I think we need to expose this for the java users to work, while still getting nice inference for scala

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/Aggregator.scala [hunk 4]
```diff
@@ -47,7 +47,7 @@ import org.apache.spark.sql.{Dataset, DataFrame, TypedColumn}
  * @tparam B The type of the intermediate value of the reduction.
  * @tparam C The type of the final result.
  */
-abstract class Aggregator[-A, B, C] {
+abstract class Aggregator[-A, B, C] extends Serializable {
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/Aggregator.scala]: good idea

File: sql/core/src/main/scala/org/apache/spark/sql/GroupedDataset.scala (hunk 3 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/GroupedDataset.scala]: We need to provide a `tEncoder` here, which is not the case of `GroupedData`.

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/GroupedDataset.scala]: Okay, we still need to add scaladoc then.

**Comment by cloud-fan** [comment]
retest this please.

**Comment by marmbrus** [comment]
Thanks!  Merging to master and 1.6.

**Pull request closed without merging**